src/problems/basic_problems.jl

```diff
@@ -364,30 +354,58 @@ compile time whether the integrator function is in-place.
 
 The fields match the names of the constructor arguments.
 """
-struct IntegralProblem{isinplace, P, F, B, K} <: AbstractIntegralProblem{isinplace}
+struct IntegralProblem{isinplace, P, F, T, K} <: AbstractIntegralProblem{isinplace}
     f::F
-    lb::B
-    ub::B
-    nout::Int
+    domain::T
     p::P
-    batch::Int
     kwargs::K
-    @add_kwonly function IntegralProblem{iip}(f, lb, ub, p = NullParameters();
-        nout = 1,
-        batch = 0, kwargs...) where {iip}
-        @assert typeof(lb)==typeof(ub) "Type of lower and upper bound must match"
+    @add_kwonly function IntegralProblem{iip}(f::AbstractIntegralFunction{iip}, domain,
+        p = NullParameters();
+        kwargs...) where {iip}
         warn_paramtype(p)
-        new{iip, typeof(p), typeof(f), typeof(lb), typeof(kwargs)}(f,
-            lb, ub, nout, p,
-            batch, kwargs)
+        new{iip, typeof(p), typeof(f), typeof(domain), typeof(kwargs)}(f,
+            domain, p, kwargs)
     end
 end
 
 TruncatedStacktraces.@truncate_stacktrace IntegralProblem 1 4
 
-function IntegralProblem(f, lb, ub, args...;
+function IntegralProblem(f::AbstractIntegralFunction,
+    domain,
+    p = NullParameters();
     kwargs...)
-    IntegralProblem{isinplace(f, 3)}(f, lb, ub, args...; kwargs...)
+    IntegralProblem{isinplace(f)}(f, domain, p; kwargs...)
+end
+
+function IntegralProblem(f::AbstractIntegralFunction,
+    lb::B,
+    ub::B,
+    p = NullParameters();
+    kwargs...) where {B}
+    IntegralProblem(f, (lb, ub), p; kwargs...)
+end
+
+# deprecation methods, which assume integrands return Float64 values (same as C libraries)
+function IntegralProblem{iip}(f, args...; nout = 1, batch = 0, kwargs...) where {iip}
+    @warn "`nout` and `batch` keywords are deprecated in favor of inplace `IntegralFunction`s or `BatchIntegralFunction`s"
+    g = if iip
+        output_prototype = Vector{Float64}(undef, nout)
+        if batch == 0
+            IntegralFunction(f, output_prototype)
+        else
+            BatchIntegralFunction(f, output_prototype, max_batch=batch)
+        end
+    else
+        if batch == 0
+            IntegralFunction(f)
+        else
+            BatchIntegralFunction(f, max_batch=batch)
+        end
+    end
+    IntegralProblem(g, args...; kwargs...)
+end
+function IntegralProblem(f, args...; kwargs...)
+    IntegralProblem{isinplace(f, 3)}(f, args...; kwargs...)
```

Comment:
```suggestion
function IntegralProblem(f, args...; nout = nothing, batch = nothing, kwargs...)
    if nout !== nothing || batch !== nothing
       @warn "`nout` and `batch` keywords are deprecated in favor of inplace `IntegralFunction`s or `BatchIntegralFunction`s. See the updated Integrals.jl documentation for details."
    end
    
    iip = isinplace(f, 3)
    g = if iip
        nout = nout === nothing ? 1 : nout
        output_prototype = Vector{Float64}(undef, nout)
        if batch == 0
            IntegralFunction(f, output_prototype)
        else
            BatchIntegralFunction(f, output_prototype, max_batch=batch)
        end
    else
        if batch == 0
            IntegralFunction(f)
        else
            BatchIntegralFunction(f, max_batch=batch)
        end
    end
    IntegralProblem{iip}(g, args...; kwargs...)
```

The previous approach would always throw the warning even when those keyword arguments weren't passed.